packages/dev/core/src/AudioV2/abstractAudio/subNodes/abstractAudioSubGraph.ts

```diff
@@ -0,0 +1,144 @@
+ * of sub nodes are used at all times to save memory and CPU resources. The tradeoff is there a small delay when first
+ * setting a property backed by a sub node. This delay is avoided by using the appropriate options to initialize the
+ * sub node on creation, e.g. `spatialEnabled` and `stereoEnabled`, or by setting any creation option backed by the
+ * sub node, e.g. `spatialPosition` and `stereoPan`.
+ *
+ * @internal
+ */
+export abstract class _AbstractAudioSubGraph {
+    private _createSubNodePromises = new Map<string, Promise<_AbstractAudioSubNode>>();
+    private _subNodes = new Map<string, Set<AbstractNamedAudioNode>>();
+
+    /**
+     * Executes the given callback with the named sub node, creating the sub node if needed.
+     *
+     * Note that `callback` is executed synchronously if the sub node already exists, otherwise the sub node is created
+     * asynchronously before `callback` is executed.
+     *
+     * @param name The name of the sub node
+     * @param callback The function to call with the named sub node
+     */
+    public callOnSubNode<T extends _AbstractAudioSubNode>(name: string, callback: (node: T) => void): void {
+        const node = this.getSubNode(name);
+        if (node) {
+            callback(node as T);
+            return;
+        }
+
+        const promise = this._createSubNodePromises.get(name) ?? this._createAndAddSubNode(name);
+
+        promise?.then((node) => {
+            callback(node as T);
+        });
+    }
+
+    /**
+     * Releases associated resources.
+     */
+    public dispose() {
+        const subNodeIterator = this._subNodes.values();
+        for (let nextSubNode = subNodeIterator.next(); !nextSubNode.done; nextSubNode = subNodeIterator.next()) {
+            const subNodeSetIterator = nextSubNode.value.values();
+            for (let nextSubNodeSet = subNodeSetIterator.next(); !nextSubNodeSet.done; nextSubNodeSet = subNodeSetIterator.next()) {
+                nextSubNodeSet.value.dispose();
+            }
+        }
+
+        this._subNodes.clear();
+        this._createSubNodePromises.clear();
+    }
+
+    /**
+     * Gets a previously created sub node.
+     * @param name - The name of the sub node
+     * @returns The named sub node, or `null` if has not been created, yet
+     * @internal
+     * */
+    public getSubNode<T extends AbstractNamedAudioNode>(name: string): Nullable<T> {
+        const set = this._subNodes.get(name);
+
+        if (!set) {
+            return null;
+        }
+
+        return set.values().next().value as T;
+    }
+
+    protected abstract _createSubNode(name: string): Nullable<Promise<_AbstractAudioSubNode>>;
+
+    /**
+     * Called when sub-nodes are added or removed.
+     * - Override this to connect and reconnect sub-nodes as needed.
+     */
+    protected _onSubNodesChanged(): void {}
+
+    protected async _createSubNodePromisesResolved(): Promise<void> {
+        await Promise.all(this._createSubNodePromises.values());
```

Comment:
as you anyhow need to await it when calling the function, there is no need to add `await` here. It also doesn't have to be an `async` function. You can just return the Promise.all call directly